### PR TITLE
Implementation of the ChaCha stream cipher

### PR DESCRIPTION
**Pull request by rfswarm**
This implementation passed all tests on https://github.com/secworks/chacha_testvectors. I am currently implementing the unit tests and ask myself how many test cases we should include. Any recommendations?

### CODE REVIEW

File: sys/crypto/chacha.c [hunk 1]
```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin, Computer Systems & Telematics
```
Comment (LudwigOrtmann) [inline sys/crypto/chacha.c]: Are you employed at Freie Universität?

Comment (rfswarm) [inline sys/crypto/chacha.c]: no and I am not interrested in the copyright at all. do I need to add one?
(when I did some code based on older or already written code - i ll just leave it yes?)

Comment (OlegHahm) [inline sys/crypto/chacha.c]: > no and I am not interrested in the copyright at all. do I need to add one?

Concerning German law this line is pretty useless anyway - your the author and therefore copyright holder. For US other foreign laws, I don't know what happens if you omit this line. If it doesn't bother you, I think it's easiest if you just put your name here. (Of course, you can also grant it also to someone else - or even to the university. At least for US law.)

> (when I did some code based on older or already written code - i ll just leave it yes?)

Yes, but you are allowed to add your own copyright.

Comment (OlegHahm) [inline sys/crypto/chacha.c]: Do you plan to change this or are you granting the copyright to FU Berlin?

File: sys/crypto/chacha.c [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_crypto 
+ * @{
+ *
+ * @file        chacha.c
```
Comment (OlegHahm) [inline sys/crypto/chacha.c]: Remove the parameter for `@file`. Doxygen can handle this better.

File: sys/crypto/chacha.c [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+#include <stdio.h>
+#include <stdint.h>
+
+#include "crypto/chacha.h"
+
+/* Perform rounds/2 number of doublerounds */
+void doublerounds(uint8_t output[64], const uint32_t input[16], uint8_t rounds)
+{
+  uint32_t x[16];
```
Comment (OlegHahm) [inline sys/crypto/chacha.c]: Coding conventions -> indentation by 4 white spaces

File: sys/include/crypto/chacha.h [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_crypto
+ * @{
+ *
+ * @file        chacha.h
```
Comment (OlegHahm) [inline sys/include/crypto/chacha.h]: see above

**Comment by OlegHahm** [comment]
```
ERROR: The following new files generate doxygen warnings:

sys/include/crypto/chacha.h
```

**Comment by OlegHahm** [comment]
@rfswarm, ping!

**Comment by Kijewski** [comment]
Currently I am in favor of closing this PR, as the copyright / authorship claim is dubious.

**Comment by OlegHahm** [comment]
What leads to this conclusion?

**Comment by Kijewski** [comment]
The file appears to be an extract of https://github.com/secworks/chacha_testvectors/blob/18d818d97b218eac3f42db4d06e57bafe0a48234/src/gen_chacha_testvectors.c sans copyright header.

**Comment by rfswarm** [comment]
.... dude aggain ref implementation of supercop...

**Comment by OlegHahm** [comment]
This is superseded by #3229 and can be closed, right?

**Pull request closed without merging**